tools/noescape/noescape.go

```diff
@@ -0,0 +1,396 @@
+// Copyright 2019 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package noescape allows recursive escape analysis for hot paths.
+//
+// To the use the package, annotation a function-level comment with either the
+// line "// +noescape" or "// +noescape:OPTION[,OPTION]". In the second case,
+// only the given types will be checked. Valid types include:
+//
+// 	interface: A call is made via an interface.
+// 	heap: A direct allocation is made on the heap.
+// 	builtin: A call is made to a built-in allocation function.
+// 	closure: A closure is allocated.
+// 	dynamic: A dynamic function is dispatched (similar to an interface).
+//
+// The key 'local' may also be specified as an OPTION. This means that only
+// local escapes are considered: escapes in subcalls are allowed.
+//
+// The key 'all' may also be specified as an OPTION. This includes all reasons,
+// local to the function and in invoked functions. Similarly, the key
+// 'explicit' may be specified, which limits the scope to hard escapes only.
```

Comment:
Is there a description somewhere of the difference between hard and soft escapes?

Reply:
They are runtime internals. Unfortunately, I don't think there's a better way to encode this, but they are unlikely to change. I need to test more.